Add device_status and action_pending to /hosts API

kc9wwh opened the pull request:
<!-- Add the related story/sub-task/bug number, like Resolves #123, or remove if NA -->
**Related issue:** Resolves #36315 


## Testing

- [X] Added/updated automated tests
- [X] Where appropriate, [automated tests simulate multiple hosts and test for host isolation](https://github.com/fleetdm/fleet/blob/main/docs/Contributing/reference/patterns-backend.md#unit-testing) (updates to one hosts's records do not affect another)
- [ ] QA'd all new/changed functionality manually
  - @kc9wwh: No issues with building or enrolling devices. Got it behind SSL cert and MDM working, but I would need a supervised device to test further.

claude committed:
Add device_status and pending_action to ListHosts API endpoint

Implements issue #34923 to add device_status and pending_action fields
to the /api/v1/fleet/hosts endpoint, allowing bulk retrieval of lock/wipe
status for all hosts.

Implementation uses a batch query approach to maintain performance:
- Added GetHostsLockWipeStatusBatch() datastore method that efficiently
  queries lock/wipe status for multiple hosts in ~5-10 queries total
- Batch queries host_mdm_actions, MDM commands (Apple/Windows), and
  script results for all hosts at once
- Reuses existing HostLockWipeStatus logic to ensure consistency
  with single-host endpoint

Changes:
- server/datastore/mysql/scripts.go: Implemented batch query method
- server/fleet/datastore.go: Added interface definition
- server/service/hosts.go: Updated ListHosts to populate device_status
  and pending_action for all hosts
- server/mock/datastore_mock.go: Added mock implementation
- server/service/hosts_test.go: Added comprehensive unit tests
- server/datastore/mysql/hosts_test.go: Added integration tests
- docs/REST API/rest-api.md: Updated API documentation

No schema changes, no dependency changes - works with existing database.

Addresses issue #34923

claude committed:
Fix type mismatch with MDMHostData

Fixed compilation errors where host.MDM was incorrectly compared to nil.
The MDM field is of type MDMHostData (a struct, not a pointer), so it
cannot be compared to nil. Removed the nil check since the struct is
always present.

Also fixed test code to use fleet.MDMHostData{} instead of
&fleet.HostMDM{} which was causing a type mismatch.

Changes:
- server/service/hosts.go: Removed invalid nil check for host.MDM
- server/service/hosts_test.go: Fixed MDM field initialization to use
  correct type (MDMHostData instead of *HostMDM)

kc9wwh committed:
Fix device status and pending action logic for MDM hosts

kc9wwh commented:
Ran osquery-perf: 
```
2025/11/19 16:20:22 uptime: 22m0s, error rate: 0.00, osquery enrolls: 9056, orbit enrolls: 4563, mdm enrolls: 0, distributed/reads: 135468, distributed/writes: 9056, config requests: 13249, result log requests: 30403, mdm sessions initiated: 0, mdm commands received: 0, config errors: 0, distributed/read errors: 0, distributed/write errors: 0, log result errors: 0, orbit errors: 0, desktop errors: 10703, mdm errors: 0, ddm declaration items success: 0, ddm declaration items errors: 0, ddm activation success: 0, ddm activation errors: 0, ddm configuration success: 0, ddm configuration errors: 0, ddm status success: 0, ddm status errors: 0, buffered logs: 0, script execs (errs): 0 (0), software installs (errs): 0 (0)
```

Hitting the `/hosts` endpoint:
```
└ $ curl -k -H "Authorization: Bearer $TOKEN" \
  "https://localhost:8080/api/v1/fleet/hosts" | jq '.hosts[] | {
    id: .id,
    hostname: .hostname,
    mdm: {
      enrollment_status: .mdm.enrollment_status,
      device_status: .mdm.device_status,
      pending_action: .mdm.pending_action
    }
  }'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 20.0M    0 20.0M    0     0  36.6M      0 --:--:-- --:--:-- --:--:-- 36.5M
{
  "id": 1,
  "hostname": "cHZbvu89KQeb",
  "mdm": {
    "enrollment_status": null,
    "device_status": "unlocked",
    "pending_action": ""
  }
}
{
  "id": 2,
  "hostname": "Mt8pTqetZ2.N",
  "mdm": {
    "enrollment_status": "Off",
    "device_status": "unlocked",
    "pending_action": ""
  }
}
{
  "id": 3,
  "hostname": "XTmdZ2V7zoq6",
  "mdm": {
    "enrollment_status": "Off",
    "device_status": "unlocked",
    "pending_action": ""
  }
}
{
  "id": 4,
  "hostname": "P0pKXSpIWU-P",
  "mdm": {
    "enrollment_status": "Off",
    "device_status": "unlocked",
    "pending_action": ""
  }
}
{
  "id": 5,
  "hostname": "CW4Pz9ZF6xH6",
  "mdm": {
    "enrollment_status": "Off",
    "device_status": "unlocked",
    "pending_action": ""
  }
}
...
```

JordanMontgomery commented on server/datastore/mysql/scripts.go:
```diff
@@ -1498,6 +1498,438 @@ func (ds *Datastore) GetHostLockWipeStatus(ctx context.Context, host *fleet.Host
+	}
+
+	// Build map of host ID to host for lookups
+	hostMap := make(map[uint]*fleet.Host, len(hosts))
```
This map is not atually used as far as I can tell

JordanMontgomery reviewed (commented):
A few tests appear to be failing as well as lint(which looks like a minor one line fix). Also I think this PR could potentially use an integration test update but we can handle that as part of the engineering side of testing/implementation if needed

kc9wwh committed:
removed `hostMap`

kc9wwh committed:
- Fix lint error: remove unnecessary zero initialization in
hosts_test.go - Add device_status and pending_action population to
ListHostsInLabel endpoint - Update comment in
GetHostsLockWipeStatusBatch for accuracy

kc9wwh committed:
Added mock for `GetHostsLockWipeStatusBatch` to the `TestListHosts`
function

JordanMontgomery committed:
Remove REST API changes

kc9wwh committed:
Added missing checksum field to `hosts_test.go`, removed incorrect
`IsLocked()` logice from `scrips.go` and updated service unit tests.

kc9wwh committed:
Added missing scope and mock function

kc9wwh committed:
Fixed `nano_command_results` insert to use valid XML and added
`GetHostsLockWipeStatusBatchFunc` mock to TestGetHostsMDM

kc9wwh committed:
Added `nano_devices` and `nano_enrollments` setup for foreign key
constraint

kc9wwh committed:
Fixed MySQL test expectation to match integration test behavior

kc9wwh committed:
Created proper Windows enrollment and response records for Windows MDM
command results

kc9wwh committed:
Fixed Windows command result mapping by joining with enrollments table
to get host_uuid

kc9wwh commented:
@JordanMontgomery all checks now passing. 🖖🏻

MagnusHJensen committed:
gate device status and pending action behind list option param

MagnusHJensen committed:
share svc and ds methods for batch and non batch

MagnusHJensen committed:
don't share methods

MagnusHJensen committed:
fix return

MagnusHJensen committed:
parse url param and add case to integration test

MagnusHJensen committed:
test labels in host matches with populate_device_status

JordanMontgomery committed:
Merge branch 'main' into kc9wwh-hosts-api-status-action_pending-device_status
